core/src/avm2/array.rs

```diff
@@ -11,8 +14,229 @@ use std::{cmp::max, ops::RangeBounds};
 /// the prototype.
 #[derive(Clone, Collect)]
 #[collect(no_drop)]
-pub struct ArrayStorage<'gc> {
-    storage: Vec<Option<Value<'gc>>>,
+pub enum ArrayStorage<'gc> {
+    Dense(Vec<Option<Value<'gc>>>, usize),
+    Sparse(BTreeMap<usize, Value<'gc>>, usize),
+}
+
+struct ArrayStorageIterator<'a, 'gc> {
+    storage: &'a ArrayStorage<'gc>,
+    index: usize,
+    index_back: usize,
+}
+
+pub struct ArrayStorageMutableIterator<'a, 'gc> {
+    storage: &'a mut ArrayStorage<'gc>,
+    index: usize,
+    index_back: usize,
+}
+
+impl<'a, 'gc> Iterator for ArrayStorageIterator<'a, 'gc> {
+    type Item = Option<Value<'gc>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match &self.storage {
+            ArrayStorage::Dense(storage, _) => {
+                if self.index >= self.index_back {
```

Comment:
Since this check is common to both branches, can it be moved outside the `match`?